driver/xrt/include/fpgabuffer.hpp

```diff
@@ -0,0 +1,147 @@
+  FPGABuffer(dtype *buffer, addr_t length, dataType type, xrt::device &device,
+             xrt::memory_group mem_grp)
+      : Buffer<dtype>(nullptr, length, type, 0x0),
+        bo(device, get_aligned_buffer(buffer, length * sizeof(dtype)),
+           length * sizeof(dtype), mem_grp) {
+    set_buffer();
+  }
+
+  FPGABuffer(addr_t length, dataType type, xrt::device &device,
+             xrt::memory_group mem_grp)
+      : Buffer<dtype>(nullptr, length, type, 0x0), is_aligned(true),
+        bo(device, length * sizeof(dtype), mem_grp) {
+    set_buffer();
+    // Initialize memory to zero
+    memset(this->_buffer, 0, this->_size);
+  }
+
+  // copy constructor
+  FPGABuffer(xrt::bo bo_, addr_t length, dataType type, bool is_aligned_,
+             dtype *unaligned_buffer_)
+      : Buffer<dtype>(nullptr, length, type, 0x0), is_aligned(is_aligned_),
+        bo(bo_), aligned_buffer(bo.map<dtype *>()),
+        unaligned_buffer(unaligned_buffer_) {
+    set_buffer();
+  }
+
+  virtual ~FPGABuffer() {
+    if (!is_aligned && own_unaligned) {
+#if (__cplusplus >= 201703L)
+      std::free(aligned_buffer);
+#else
+      free(aligned_buffer);
+#endif
+    }
+  }
+
+  void sync_from_device() override {
+    bo.sync(xclBOSyncDirection::XCL_BO_SYNC_BO_FROM_DEVICE);
+    if (!is_aligned) {
+      memcpy(unaligned_buffer, aligned_buffer, this->size());
+    }
+  }
+
+  void sync_to_device() override {
+    if (!is_aligned) {
+      memcpy(aligned_buffer, unaligned_buffer, this->size());
+    }
+    bo.sync(xclBOSyncDirection::XCL_BO_SYNC_BO_TO_DEVICE);
+  }
+
+  void free_buffer() override { return; }
+
+  std::unique_ptr<BaseBuffer> slice(size_t start, size_t end) override {
+    size_t start_bytes = start * sizeof(dtype);
+    size_t end_bytes = end * sizeof(dtype);
+
+    dtype *offset_unaligned_buffer = nullptr;
+    if (!is_aligned) {
+      offset_unaligned_buffer = &unaligned_buffer[start];
+    }
+
+    return std::unique_ptr<BaseBuffer>(new FPGABuffer(
+        xrt::bo(bo, end_bytes - start_bytes, start_bytes), end - start,
+        this->_type, this->is_aligned, offset_unaligned_buffer));
+  }
+
+private:
+  bool is_aligned;
+  bool own_unaligned{};
+  xrt::bo bo;
+  dtype *aligned_buffer;
+  dtype *unaligned_buffer;
+
+  dtype *get_aligned_buffer(dtype *host_buffer, size_t length) {
+    if ((reinterpret_cast<uintptr_t>(host_buffer) % ALIGNMENT) != 0) {
+      std::cerr
+          << "[ACCL] Warning: you're creating a buffer from an unaligned host "
+             "pointer. This will cause extra copying when syncing the buffers."
+          << std::endl;
+      size_t aligned_size =
+          ((size_t)ceil(length * sizeof(dtype) / (double)ALIGNMENT)) *
+          ALIGNMENT;
+      is_aligned = false;
+#if (__cplusplus >= 201703L)
+      aligned_buffer =
+          static_cast<dtype *>(std::aligned_alloc(ALIGNMENT, aligned_size));
+#else
+      dtype *aligned_buffer;
+      posix_memalign(static_cast<void **>(&aligned_buffer), ALIGNMENT,
```

Comment:
We need to reinterpret this pointer to make this work.

```suggestion
      posix_memalign(reinterpret_cast<void **>(&aligned_buffer), ALIGNMENT,
```
```suggestion
      posix_memalign(static_cast<void **>(&aligned_buffer), ALIGNMENT,
```